### PR TITLE
feat: add popular network list modal

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**

This PR introduces the addition of a Popular Networks list to the Add Network modal.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25160?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Run `yarn && ENABLE_NETWORK_UI_REDESIGN=1 yarn start`
2. Go to Settings -> Developer Options
3. Tun on the network new toggle
4. Go to the wallet page
5. Click on the network button ( see the video below )
6. you should see the list of popular network
7. you should be able to add a network from the modal

### **Before**

https://github.com/MetaMask/metamask-extension/assets/26223211/bde545ef-c5a3-4cc5-9508-16edfd8507dd

### **After**

https://github.com/MetaMask/metamask-extension/assets/26223211/114076c2-e336-4b96-b061-defd6d886380

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                  );
+                }}
+              >
+                {t('add')}
+              </Button>
+            </Box>
+          </Box>
+        ))}
+      </Box>
+      <Box padding={isPopUp ? [2, 0, 2, 6] : [2, 0, 2, 0]}></Box>
```
Comment (darkwing) [inline ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx:186]: Do we need this empty `<Box>`?

File: ui/pages/routes/routes.component.js [hunk 2]
```diff
@@ -928,6 +929,9 @@ export default class Routes extends Component {
         {isNetworkMenuOpen ? (
           <NetworkListMenu onClose={() => toggleNetworkMenu()} />
         ) : null}
+        {process.env.ENABLE_NETWORK_UI_REDESIGN ? (
+          <NetworkConfirmationPopover />
```
Comment (darkwing) [inline ui/pages/routes/routes.component.js:933]: Can this be added in `network-list-menu.js` instead of `Routes`?

File: ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                key="link"
+                className="add-network__edge-case-box__link"
+                href="https://chainlist.wtf/"
+                target="_blank"
+                rel="noreferrer"
+              >
+                {t('here')}.
+              </a>,
+              <ButtonPrimary
+                // backgroundColor={BackgroundColor.backgroundDefault}
```
Comment (darkwing) [inline ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx:84]: Needs removal.

**Review by darkwing — APPROVED** [review]
Functionally this works great!

**Review by darkwing — CHANGES_REQUESTED** [review]
Sorry, requested a few changes

File: ui/components/multichain/network-list-menu/network-confirmation-popover/network-confirmation-popover.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const unapprovedConfirmations = useSelector(getUnapprovedConfirmations);
+
+  useEffect(() => {
+    const anAddNetworkConfirmationFromMetaMaskExists =
+      unapprovedConfirmations?.find(
+        (confirmation) =>
+          confirmation.origin === 'metamask' &&
+          confirmation.type === ApprovalType.AddEthereumChain,
+      );
+    console.log('HERE ====>', unapprovedConfirmations);
```
[the remark this answers is not in the document]
Comment (salimtb) [reply ui/components/multichain/network-list-menu/network-confirmation-popover/network-confirmation-popover.tsx:20]: ooh nice , thanks for the tip

File: ui/pages/routes/routes.component.js (hunk 2 above)
Comment (salimtb) [reply ui/pages/routes/routes.component.js:933]: The issue with incorporating this into network-list-menu.js lies in its nature as a modal component. For the functionality in question, utilizing a popover necessitates it being part of a standard component that is associated with a specific route, rather than being embedded within a modal.

File: ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+              alignItems={AlignItems.center}
+              marginLeft={1}
+            >
+              <Button
+                type={ButtonVariant.Link}
+                className="add-network__add-button"
+                variant={ButtonVariant.Link}
+                data-testid="test-add-button"
+                onClick={async () => {
+                  dispatch(toggleNetworkMenu());
```
Comment (darkwing) [inline ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx:160]: Let's do this call second so that it disappears in the background instead of before the `requestUserApproval` finishes.

File: ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  searchAddNetworkResults,
+}: {
+  searchAddNetworkResults: RPCDefinition[];
+}) => {
+  const t = useI18nContext();
+  const isPopUp = getEnvironmentType() === ENVIRONMENT_TYPE_POPUP;
+  const dispatch = useDispatch();
+  const history = useHistory();
+
+  if (Object.keys(searchAddNetworkResults).length === 0) {
```
Comment (darkwing) [inline ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx:51]: If the user has already added all popular networks, can we just not show that list?

File: ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx (hunk 5 above)
Comment (salimtb) [reply ui/components/multichain/network-list-menu/popular-network-list/popular-network-list.tsx:160]: eeem i think we will have a problem if we do it that way, The issue arises because `requestUserApproval` is an asynchronous function. When we use `await requestUserApproval()`, the execution pauses, waiting for the user to either approve or reject through the confirmation modal. This means the promise stays in a pending state until the user interacts with the "add network" confirmation modal, which is triggered by `requestUserApproval`. However, there's a complication: our modal ends up obscuring the confirmation modal because `dispatch(toggleNetworkMenu());` is called after, not before, this interaction. As a result, from a functional standpoint, we encounter a problem where the user might not be able to see or interact with the confirmation modal as intended.

**Review by darkwing — APPROVED** [review]

**Review by ustad56 — APPROVED** [review]

**Review by bergeron — APPROVED** [review]
Worked for me locally, looks good

File: ui/components/multichain/network-list-menu/network-list-menu.js [hunk 7]
```diff
@@ -95,6 +105,17 @@ export const NetworkListMenu = ({ onClose }) => {
...
+    (net) => net.chainId,
+  );
+
+  const sortedFeaturedNetworks = FEATURED_RPCS.sort((a, b) =>
+    a.nickname > b.nickname ? 1 : -1,
+  ).slice(0, FEATURED_RPCS.length);
+
+  const notExistingNetworkConfigurations = sortedFeaturedNetworks.filter(
+    ({ chainId }) => !networkConfigurationChainIds.includes(chainId),
+  );
```
Comment (bergeron) [inline ui/components/multichain/network-list-menu/network-list-menu.js:118]: maybe stuff that could be memoized with `useMemo` to avoid recalculating when unrelated stuff changes.  For example, dragging networks to a different order will re-render this  component, but shouldn't change this calculation.
